logics/impls/controller/mod.rs

```diff
@@ -1516,83 +1293,119 @@ impl<T: Storage<Data>> Internal for T {
             }
 
+            let borrow_balance_stored = pool_attribute.account_borrow_balance;
             if borrow_balance_stored != 0 {
-                let borrow_balance_eth = U256::from(unit_price)
+                let borrow_balance_eth = U256::from(oracle_price)
                     .mul(U256::from(borrow_balance_stored))
                     .div(U256::from(PRICE_PRECISION));
                 total_debt_in_base_currency = total_debt_in_base_currency.add(borrow_balance_eth);
             }
 
-            attr_pool
+            asset_params.push(HypotheticalAccountLiquidityCalculationParam {
+                asset: attr_pool,
+                decimals: pool_attribute.decimals,
+                token_balance: pool_attribute.account_balance,
+                borrow_balance: pool_attribute.account_borrow_balance,
+                exchange_rate_mantissa: Exp {
+                    mantissa: WrappedU256::from(pool_attribute.exchange_rate),
+                },
+                collateral_factor_mantissa: Exp {
+                    mantissa: collateral_factor_mantissa,
+                },
+                oracle_price_mantissa: oracle_price_mantissa.clone(),
+            });
+
+            (
+                oracle_price,
+                pool_attribute.liquidation_threshold,
+                attr_pool,
+            )
         } else {
-            caller
+            (0, 0, caller)
         };
 
+        // NOTE: Do not use account_assets as it makes doubled cross-contract calling leads to high gas.
         for asset in self._markets() {
-            if asset == skip_pool || asset == caller {
+            if asset == skip_pool {
                 continue
             }
-            let (compounded_liquidity_balance, borrow_balance_stored, _) =
+            // Read the balances and exchange rate from the pool
+            let (compounded_liquidity_balance, borrow_balance_stored, exchange_rate_mantissa) =
                 PoolRef::get_account_snapshot(&asset, account);
+
             if compounded_liquidity_balance == 0 && borrow_balance_stored == 0 {
                 continue
             }
 
-            let collateral_factor_mantissa: WrappedU256 = self
+            let decimals = PoolRef::token_decimals(&asset);
+            // Get the normalized price of the asset
+            let oracle_price: u128 =
+                PriceOracleRef::get_underlying_price(&oracle, asset).ok_or(Error::PriceError)?;
+            if oracle_price == 0 {
+                return Err(Error::PriceError)
+            }
+            let oracle_price_mantissa = Exp {
+                mantissa: WrappedU256::from(U256::from(oracle_price)),
+            };
+
+            let collateral_factor_mantissa = self
                 ._collateral_factor_mantissa(asset)
                 .ok_or(Error::MarketNotListed)?;
             let ltv = U256::from(collateral_factor_mantissa);
 
             let liquidation_threshold = PoolRef::liquidation_threshold(&asset);
```

Comment:
[WANT]
For gas reduction, it may be a good idea to put together a function to get the metadata of a pool.
(Further optimization will increase the amount of changes, so this seems to be all we need for now.)

```rust
PoolRef::token_decimals(&asset)
PoolRef::liquidation_threshold(&asset);
```

Reply:
I don't want to put it in the scope right now,  but in the future, I think it would be good to organize it into a function that can get all the information from the Pool at once, since I think there will be a lot of gas in the Pool calls
For example
[1] PoolRef::metadata
Get pool metadata (underlying, decimals, liquidation_threshold etc.)
[2] PoolRef::status
get (potentially variable) status of the pool (total supply, exchange rate, etc.)
[3] PoolRef::account_snapshot
information per account (≈ get_account_snapshot)